library/core/src/iter/adapters/skip.rs

```diff
@@ -33,21 +33,27 @@ where
     #[inline]
     fn next(&mut self) -> Option<I::Item> {
         if unlikely(self.n > 0) {
-            self.iter.nth(crate::mem::take(&mut self.n) - 1);
+            self.iter.nth(crate::mem::take(&mut self.n))
+        } else {
+            self.iter.next()
         }
-        self.iter.next()
     }
 
     #[inline]
     fn nth(&mut self, n: usize) -> Option<I::Item> {
-        // Can't just add n + self.n due to overflow.
         if self.n > 0 {
-            let to_skip = self.n;
+            // Checked add to handle overflow case.
+            let n = self.n.checked_add(n).unwrap_or_else(|| {
+                // In case of overflow, load skip value, then load `n`.
+                self.iter.nth(self.n);
```

Comment:
This feels off to me -- previously the logic had the equivalent of nth(self.n - 1), nth(n) here, I think? But now we're doing nth(self.n) and then nth(n), which is off by one.

Do we have tests for this function? I'm a little surprised this didn't trigger any change in them.

Reply:
I'm not sure exactly how this should be tested. This PR primarily intends to reduce the amount of unnecessary intermediate loading of values, but this one can't necessarily be avoided because some value has to be loaded in-between 0..n on overflow to make sure the iterator is fully consumed but does not return the value at `usize::MAX`.

The index was off by one, but a test would have to be written to check that the intermediate value discarded was at the "right" place, is this valuable? I'll just open the PR and someone can comment if tests should be added in a specific way.

Reply:
My impression is that the index being off by one means that an iterator would return a *different* element. For example, an iterator over (0..=u128::MAX).skip(usize::MAX).nth(usize::MAX) should return u128::MAX, but I think with the previous code would have returned u128::MAX-1.

Can you add a test for this?

Reply:
yeah, my apologies, I had the -1 logic added, but I accidentally deleted it when applying the other fix. https://github.com/rust-lang/rust/pull/96350/commits/f1f5e08378f4ceb52992f5d9c3abb009bab7774f should have this change. The check you mentioned shouldn't return `u128::MAX`, I don't believe, but I wrote a test that does.